pkg/reconciler/labeler/v2/accessors.go

```diff
@@ -106,40 +107,52 @@ func markRoutingState(
 	}
 }
 
-// addRouteLabel appends the route label to the list of labels if needed
-// or removes the label if routeName is nil.
-func addRouteLabel(acc kmeta.Accessor, routeName string, diffLabels map[string]interface{}) error {
-	if routeName == "" { // remove the label
-		if acc.GetLabels()[serving.RouteLabelKey] != "" {
-			diffLabels[serving.RouteLabelKey] = nil
+// updateRouteAnnotation appends the route annotation to the list of labels if needed
+// or removes the annotation if routeName is nil.
+// Returns true if the entire annotation is newly added or removed, which signifies a state change.
+func updateRouteAnnotation(acc kmeta.Accessor, routeName string, diffAnn map[string]interface{}, remove bool) bool {
+	valSet := getListAnnValue(acc.GetAnnotations(), serving.RoutesAnnotationKey)
+	has := valSet.Has(routeName)
+	switch {
+	case has && remove:
+		if len(valSet) == 1 {
+			diffAnn[serving.RoutesAnnotationKey] = nil
+			return true
 		}
-	} else { // add the label
-		if oldLabel := acc.GetLabels()[serving.RouteLabelKey]; oldLabel == "" {
-			diffLabels[serving.RouteLabelKey] = routeName
-		} else if oldLabel != routeName {
-			// TODO(whaught): this restricts us to only one route -> revision
-			// We can move this to a comma separated list annotation and use the new routingState label.
-			return fmt.Errorf("resource already has route label %q, and cannot be referenced by %q", oldLabel, routeName)
+		valSet.Delete(routeName)
+		diffAnn[serving.RoutesAnnotationKey] = strings.Join(valSet.List(), ",")
```

Comment:
Do we need `List` here that sorts things or `UnsortedList` will work?

Reply:
UnsortedList was unstable and sometimes failed the unit test expectation. Sorted makes it easier to know what to expect for the annotation value.

Reply:
I know that it's random, but I'd also not rather incur additional work in the production code for UT :-) (i.e. we can split and compare sets in the test).